server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java

```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
-            LOGGER.warn("Fail to copy file " + filename + " in System VM " + ip, e);
-            details = e.getMessage();
-            success = false;
-        }
-        return new ExecutionResult(success, details);
-    }*/
-
-    public File getSystemVMKeyFile() {
-        final URL url = this.getClass().getClassLoader().getResource("scripts/vm/systemvm/id_rsa.cloud");
-        File keyFile = null;
-        if (url != null) {
-            keyFile = new File(url.getPath());
-        }
-        if (keyFile == null || !keyFile.exists()) {
-            keyFile = new File("/usr/share/cloudstack-common/scripts/vm/systemvm/id_rsa.cloud");
-        }
-        assert keyFile != null;
-        if (!keyFile.exists()) {
-            LOGGER.error("Unable to locate id_rsa.cloud in your setup at " + keyFile.toString());
-        }
-        return keyFile;
-    }
-
-    private static String getRouterSshControlIp(final NetworkElementCommand cmd) {
-        final String routerIp = cmd.getAccessDetail(NetworkElementCommand.ROUTER_IP);
-        final String routerGuestIp = cmd.getAccessDetail(NetworkElementCommand.ROUTER_GUEST_IP);
-        final String zoneNetworkType = cmd.getAccessDetail(NetworkElementCommand.ZONE_NETWORK_TYPE);
-
-        if (routerGuestIp != null && zoneNetworkType != null && DataCenter.NetworkType.valueOf(zoneNetworkType) == DataCenter.NetworkType.Basic) {
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug("In Basic zone mode, use router's guest IP for SSH control. guest IP : " + routerGuestIp);
+    private boolean checkAndStartApache() {
+        //Check whether the Apache server is running
+        Script command = new Script("/bin/systemctl", LOGGER);
+        command.add("is-active");
+        command.add("apache2");
+        String result = command.execute();
+
+        //Apache Server is not running. Try to start it.
+        if (result != null && !result.equals("active")) {
+            command = new Script("/bin/systemctl", LOGGER);
+            command.add("start");
+            command.add("apache2");
+            result = command.execute();
+            if (result != null) {
+                LOGGER.warn("Error in starting apache2 service err=" + result);
+                return false;
             }
-
-            return routerGuestIp;
         }
 
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("Use router's private IP for SSH control. IP : " + routerIp);
+        return true;
+    }
+
+    protected void createFolder(String path) {
+        File dir = new File(path);
+        if (!dir.exists()) {
+            dir.mkdirs();
         }
-        return routerIp;
     }
 
-    protected Answer copyFileFromSystemVm(final NetworkElementCommand cmd, String script) throws Exception {
-        final File keyFile = getSystemVMKeyFile();
+    private CreateEntityDownloadURLAnswer downloadCompressedFileToSSVM(CreateEntityDownloadURLCommand cmd, final String diagnosticsZipFileName) {
```

Comment:
can you break this up in smaller methods?
i.e. checkApache(), mkdirCommand(), linkCommand() etc...